### PR TITLE
[android] - NPX mapZoomButtonController

### LINKED ISSUE #8326: MapBoxView does crash on visibility="gone"
**Platform:** Android
**Mapbox SDK version:** 5.0.0-beta.1

### Steps to trigger behavior

 1. Do something like
```
   <com.mapbox.mapboxsdk.maps.MapView
        android:id="@+id/mapView"
        android:layout_width="fill_parent"
        android:layout_height="fill_parent"
        android:background="@color/transparent"
        android:visibility="gone" <<<<<<<<<
      />
```

 2. Try to open that view
 3. You get a crash, with a NPE in

MapView:
```
  // Called when view is hidden and shown
  @Override
  protected void onVisibilityChanged(@NonNull View changedView, int visibility) {
    if (isInEditMode()) {
      return;
    }
    >>>>>the mapZoomButtonController is NULL
    mapZoomButtonController.setVisible(visibility == View.VISIBLE);
  }
```

### Expected behavior
View is hidden
### Actual behavior
NPE crashing

### PR DESCRIPTION
**Pull request by tobrun**
closes #8326

### CODE REVIEW

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java [hunk 1]
```diff
@@ -521,13 +521,15 @@ public void onSurfaceTextureUpdated(SurfaceTexture surface) {
   @CallSuper
   protected void onDetachedFromWindow() {
     super.onDetachedFromWindow();
-    mapZoomButtonController.setVisible(false);
+    if (mapZoomButtonController != null) {
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: What do you think of moving the check into a `private` method?
- It increases the chances that other methods can use a method when the method is finely grained.
- It allows the higher-level methods to read more like a series of comments (`public` methods should look like pseudo code, without language statements, etc.).

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java [hunk 2]
```diff
@@ -521,13 +521,15 @@ public void onSurfaceTextureUpdated(SurfaceTexture surface) {
...
+      mapZoomButtonController.setVisible(false);
+    }
   }
 
   // Called when view is hidden and shown
   @Override
   protected void onVisibilityChanged(@NonNull View changedView, int visibility) {
-    if (isInEditMode()) {
+    if (isInEditMode() || mapZoomButtonController == null) {
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: What do you think of introducing a _explaining variable_ here?
_Explaining variables_ are particularly valuable with conditional logic in which it is useful to take each clause of a condition and explain what the condition means with a well-named temp.
What about putting the result of the expression in a temp variable with a name that explains the purpose?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java (hunk 1 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: Guidelines indicate avoiding internal getters and setters: https://developer.android.com/training/articles/perf-tips.html#GettersSetters

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java (hunk 2 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: I'm not seeing the added benefit here for null checking as:

```
boolean mapboxZoomButtonControllerIsNotNull = mapbooxZoomButtonController == null;
if(isInEditMode() || mapboxZoomButtonControllerIsNotNull){
   ///
}
```

imo this is harder to read as a simple null check, happy to hear the thoughts of @mapbox/android on this.

Comment (zugaldia) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: Agree with the value of explaining variables but at the same time I also think the code is well readable as it is.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java (hunk 1 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: Also happy to hear @mapbox/android on this as well

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java (hunk 2 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: just to clarify, I also completely agree if there is more to it than just null checking but this just feels weird to write/read:
```
boolean something = somehtinToTest == null;
```

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java (hunk 1 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: If we would go this route, we should  move MapboxZoomButtonController to a seperate component.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java (hunk 2 above)
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: Expressions can become very complex and hard to read. You quickly end up with a pretty long condition. What about extracting the condition into a query?

Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: agree that they can become complex. The way how I see it, one condition ok, two condition ok, three condition -> time to apply suggestion from https://github.com/mapbox/mapbox-gl-native/pull/8332#discussion_r105248115.

Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: 👍  You follow the [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)).

Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: lol, wasn't aware of that 😅 this just felt natural to me

Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: > Never go to bed without learning something new.

👍

**Pull request merged**